### ISSUE TITLE
USDT from multiple processes

### ISSUE THREAD

**Issue by erthalion** [iovisor/bcc#2027]
If I understand correctly, current implementation doesn't support multiple processes attaching to the same USDT probe. But doing so can be really convenient in some cases, e.g. when one needs to trace PostgreSQL, since every backend is a separate process and all of them are sharing the same code. I've found few issues about this topic, namely #1106 and #1515, but not any PR. What are the plans here?

**Comment by yonghong-song** [comment]
That is right. For a probe with multiple locations, the current implementation generates code like below
```
__attribute__((always_inline))
static __always_inline int _bpf_readarg_do_trace3_1(struct pt_regs *ctx, void *dest, size_t len) {
  if (len != sizeof(int32_t)) return -1;
  switch(PT_REGS_IP(ctx)) {
  case 0x4006afULL: { u64 __addr = ctx->bp + -20; __asm__ __volatile__("": : :"memory"); int32_t __res = 0x0; bpf_probe_read(&__res, sizeof(__res), (void *)__addr); *((int32_t *)dest) = __res; } return 0;                                
  case 0x4006ceULL: *((int32_t *)dest) = ctx->ax; __asm__ __volatile__("": : :"memory"); return 0;
  }
  return -1;
}
```
Basically, for one probe, only one function is generated to extract data and different locations are filtered based on PC values. Therefore, for any probe with more than one location, pid will be required. To permit multiple processes, each location needs to have its own USDT data fill routines. 

No concrete plan yet. Anybody is interested in helping implementation on this?

cc @palmtenor

**Comment by zilder** [comment]
I also have the same problem. Not only postgres use processes for backends, but it also spawn processes to handle parallel query execution. And it is much harder to trace those because they usually live for a very short period of time.
From what I saw in the code the semaphore setup will require rework too.
